### PR TITLE
policy-correlation: enhancements and fixes to correlated policies

### PR DESCRIPTION
**Pull request by sypakine**
Several enhancements to policy correlation:

- add API kind to policy struct
- support cluster-scoped APIs
- Fix

Policy can be derived from multiple policy APIs (k8s np, cnp, ccnp, etc.). This change includes that Kind in the correlated policy.

Including the Kind in the correlated policy simplifies identifying the resource responsible for enforcement.

Release note:

```release-note
Policy correlation:
- include policy API Kind in the policy correlation result
- correlate cluster-scoped APIs (e.g. CCNP)
- correctly correlate L4 and protocol-only policies
```

### CODE REVIEW

**Comment by sypakine** [comment]
The test failures appear unrelated:
- E2E failure was from missing error log
- Conformance failure was due to pod creation

**Comment by ovidiutirla** [comment]
Re-triggered the failing tests, all the required tests are now passing.

File: api/v1/flow/flow.proto [hunk 1]
```diff
@@ -464,6 +464,7 @@ message Policy {
 	string namespace = 2;
 	repeated string labels = 3;
 	uint64 revision = 4;
+    string kind = 5;
```
Comment (gandro) [inline api/v1/flow/flow.proto]: This seems to be using spaces for indentation, where was the rest of the file uses tabs

File: pkg/policy/correlation/correlation.go [hunk 2]
```diff
@@ -134,7 +134,31 @@ func lookupPolicyForKey(ep getters.EndpointInfo, key policy.Key, matchType uint3
 	switch matchType {
 	case monitorAPI.PolicyMatchL3L4, monitorAPI.PolicyMatchL4Only:
 		// Check for L4 policy rules
-		derivedFrom, rev, ok = ep.GetRealizedPolicyRuleLabelsForKey(key)
+		if derivedFrom, rev, ok = ep.GetRealizedPolicyRuleLabelsForKey(key); ok {
+			return
+		}
+
+		// Check for allow-specific-port-protocol policies.
+		// This covers the case where one or more identities are allowed by network policy.
```
Comment (gandro) [inline pkg/policy/correlation/correlation.go]: Good catch. However, wouldn't this be the case if `matchType == monitorAPI.PolicyMatchL4Only`?

In other words, shouldn't this be it's own case under `monitorAPI.PolicyMatchL4Only`

**Review by gandro — COMMENTED** [review]
Thank you (and apologies for the late review, I was on vacation)! Looks good overall, two minor things that I think we should address however.

File: pkg/policy/correlation/correlation.go (hunk 2 above)
Comment (sypakine) [reply pkg/policy/correlation/correlation.go]: Thanks for the review. Makes sense to me Let me test that change and get back or upload the changes.

File: api/v1/flow/flow.proto (hunk 1 above)
Comment (sypakine) [reply api/v1/flow/flow.proto]: Ah, thanks. I'll push that change when I resolve the other comment.

**Comment by sypakine** [comment]
Thanks @gandro, push a new set of commits and ran the following tests to confirm:

Setup:

```
REPO_ROOT=$PWD
KUBEPROXY_MODE="none" \
WORKERS=2 \
CONTROLPLANES=1 \
CLUSTER_NAME=kind \
make kind && \
make kind-image && \
kind export kubeconfig --name kind
helm upgrade -i cilium ./install/kubernetes/cilium \
  --wait \
  --namespace kube-system \
  --set k8sServiceHost="kind-control-plane" \
  --set k8sServicePort="6443" \
  --set debug.enabled=true \
  --set debug.verbose=datapath \
  --set pprof.enabled=true \
  --set enableIPv4Masquerade=false \
  --set enableIPv6Masquerade=false \
  --set enableIPv4Masquerade=false \
  --set hostFirewall.enabled=false \
  --set socketLB.hostNamespaceOnly=true \
  --set kubeProxyReplacement=true \
  --set nodeinit.enabled=true \
  --set ipam.mode=kubernetes \
  --set ipv4.enabled=true \
  --set ipv4NativeRoutingCIDR=10.244.0.0/16 \
  --set ipv6.enabled=false \
  --set image.override="localhost:5000/cilium/cilium-dev:local" \
  --set image.pullPolicy=Never \
  --set operator.image.override="localhost:5000/cilium/operator-generic:local" \
  --set operator.image.pullPolicy=Never \
  --set operator.image.suffix="" \
  --set securityContext.privileged=true \
  --set gatewayAPI.enabled=false \
  --set socketLB.enabled=false \
  --set bpf.hostLegacyRouting=true \
  --set endpointRoutes.enabled=true \
  --set localRedirectPolicy=true

alias k="kubectl"
k apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: server
  labels:
    app: server
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchLabels:
            app: client
        topologyKey: kubernetes.io/hostname
  containers:
  - name: nginx
    image: nginx:latest
    ports:
    - containerPort: 80
---
apiVersion: v1
kind: Pod
metadata:
  name: client
  labels:
    app: client
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchLabels:
            app: server
        topologyKey: kubernetes.io/hostname
  containers:
  - name: curl
    image: curlimages/curl:latest
    command:
    - sleep
    - "9999999"
---
apiVersion: v1
kind: Service
metadata:
  name: server
spec:
  type: ClusterIP
  selector:
    app: server
  ports:
  - name: eighty
    protocol: TCP
    port: 80
    targetPort: 80
EOF
```

Confirm working w/o netpol:

```
time while ! [[ "$( k get svc server -ojsonpath='{.spec.clusterIP}' | xargs )" =~ ^10\..* ]]; do echo -n "."; sleep 1; done
clusterip=$( k get svc server -ojsonpath='{.spec.clusterIP}' )

time while ! [[ "$( k get pod server -ojsonpath='{.status.conditions[?(@.type=="Ready")].status}' )" =~ "True" ]]; do echo -n "."; sleep 1; done
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname
```

Observe flows in separate terminal:

```
alias k="kubectl"
serverhost=$( k get pod server \
    -ojsonpath='{ .spec.nodeName }' )
serveranetd=$( k get pod -n kube-system \
    -l k8s-app=cilium \
    --field-selector=spec.nodeName=${serverhost} \
    -ojsonpath='{.items[].metadata.name}' )
k exec -it -n kube-system ${serveranetd} -c cilium-agent -- hubble observe -f -ojson --type policy-verdict
```

Test each rule type:

```
// protocol-only
k apply -f - <<EOF
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: server
spec:
  podSelector:
    matchLabels:
      app: server
  ingress:
  - ports:
    - protocol: TCP
  policyTypes:
  - Ingress
EOF
sleep 5
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname


// port-only
k apply -f - <<EOF
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: server
spec:
  podSelector:
    matchLabels:
      app: server
  ingress:
  - ports:
    - port: 80
  policyTypes:
  - Ingress
EOF
sleep 5
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname

// allow-all
k apply -f - <<EOF
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: server
spec:
  podSelector:
    matchLabels:
      app: server
  ingress:
  - {}
  policyTypes:
  - Ingress
EOF
sleep 5
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname

// port+protocol (L4)
k apply -f - <<EOF
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: server
spec:
  podSelector:
    matchLabels:
      app: server
  ingress:
  - ports:
    - protocol: TCP
      port: 80
  policyTypes:
  - Ingress
EOF
sleep 5
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname

// L3+L4
k apply -f - <<EOF
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: server
spec:
  podSelector:
    matchLabels:
      app: server
  ingress:
  - from:
    - podSelector:
        matchLabels:
          app: client
    ports:
    - protocol: TCP
      port: 80
  policyTypes:
  - Ingress
EOF
sleep 5
k exec client -- curl --no-keepalive -s -m 2 -o /dev/null -v ${clusterip}:80/hostname
```

**Review by gandro — APPROVED** [review]
This is awesome and the improved testing coverage is more than welcome! Thank you so much

**Comment by gandro** [comment]
Image Build failed due to sigstore throwing a HTTP 500. Closing and reopening the PR to re-trigger CI

**Pull request merged**